encodings/fastlanes/src/bitpacking/compress.rs

```diff
@@ -236,25 +236,60 @@ where
     F: Fn(&[UnsignedT]) -> &[T],
     G: Fn(&mut [T]) -> &mut [UnsignedT],
 {
+    let my_offset_in_builder = builder.len();
+
+    builder
+        .nulls
+        .append_validity(array.validity(), array.len())?;
+
+    let packed = array.packed_slice::<UnsignedT>();
     let bit_width = array.bit_width() as usize;
     let length = array.len();
     let offset = array.offset() as usize;
+
+    unpack_values_into(
+        packed,
+        bit_width,
+        length,
+        offset,
+        builder,
+        transmute,
+        transmute_mut,
+    )?;
+
+    if let Some(patches) = array.patches() {
+        builder.patch(patches, my_offset_in_builder)?;
+    };
+
+    Ok(())
+}
+
+fn unpack_values_into<T: NativePType, UnsignedT: NativePType + BitPacking, F, G>(
+    packed: &[UnsignedT],
+    bit_width: usize,
+    length: usize,
+    offset: usize,
+    // TODO(ngates): do we want to use fastlanes alignment for this buffer?
+    builder: &mut PrimitiveBuilder<T>,
+    transmute: F,
+    transmute_mut: G,
+) -> VortexResult<()>
+where
+    F: Fn(&[UnsignedT]) -> &[T],
+    G: Fn(&mut [T]) -> &mut [UnsignedT],
+{
+    let my_offset_in_builder = builder.len();
     let last_chunk_length = match (offset + length) % 1024 {
         0 => 1024,
         last_chunk_length => last_chunk_length,
     };
 
-    builder.nulls.append_validity(array.validity(), length)?;
-
     if bit_width == 0 {
-        builder.append_zeros(length);
```

Comment:
NB: append_zeros also appends is_valids, so we end up with twice the number of validity entries.